app/models/brand.rb

```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class Brand < ApplicationRecord
+  has_many :products, dependent: :restrict_with_exception
+  validates :name, presence: true, uniqueness: true
+  def self.select_params
```

Comment:
w produkcyjnym projekcie jest duza szansa ze to by nie przeszlo. Zasmiecasz model tematami widokow. https://www.rubyguides.com/2019/09/rails-patterns-presenter-service/

Reply:
> > Rubocop mnie blokuje ze specką która sprawdza filtrowanie przez brand i category(za dużo linii), więc wrzucam bez tego, bo po refaktoringu wychodzi chyba niezbyt czytelnie:
> > ```
> >  it 'shows all products' do
> >     expect(product_cards_count(page)).to eq(10)
> >   end
> > 
> >   it 'filters products by category' do
> >     filter_by Category
> >   end
> > 
> >   it 'filters products by brand' do
> >     filter_by Brand
> >   end
> > 
> >   it 'filters products by category and brand' do
> >     products = Product.where(brand_id: Product.first.brand, category_id: Product.first.category)
> >     select Product.first.category.name, from: 'filter_category_id'
> >     select Product.first.brand.name, from: 'filter_brand_id'
> >     click_on 'Save Filter'
> >     expect(product_cards_count(page)).to eq(products.count)
> >   end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i w `spec/support/helpers.rb`
> > ```
> > module Helpers
> >   def filter_by(model)
> >     object = model.first
> >     select object.name, from: "filter_#{object.class.to_s.downcase}_id"
> >     click_on 'Save Filter'
> >     expect(product_cards_count(page)).to eq(object.products.count)
> >   end
> > 
> >   def product_cards_count(page)
> >     page.all('.product-card').count
> >   end
> > end
> > ```
> 
> Takie sytuacje beda sie zdarzac ;). Nie zawsze dasz rade spelnic rubocopa w 100%, czasami też nie warto. Jak wyglądał kod przed refactorem?

```
  it 'filters products by category' do
    category = Category.first
    select category.name, from: 'filter_category_id'
    click_on 'Save Filter'

    expect(page.all('.product-card').count).to eq(category.products.count)
  end

  it 'filters products by brand' do
    brand = Brand.first
    select brand.name, from: 'filter_brand_id'
    click_on 'Save Filter'

    expect(page.all('.product-card').count).to eq(brand.products.count)
  end

it 'filters products by category and brand' do
    product = Product.first
    brand = product.brand
    category = product.category

    products = Product.where(brand_id: brand, category_id: category)

    select category.name, from: 'filter_category_id'
    select brand.name, from: 'filter_brand_id'
    click_on 'Save Filter'

    expect(product_cards_count(page)).to eq(products.count)
  end

```

Wywalało za dużo linii w ostatim przykładzie i za dużo w całym specu.  
Żeby zmieścić się w ostatnim pozbyłem się  zmiennych `product`, `brand` i `category` 
Żeby zmieścić sie w limicie w specu wyekstraktowałem 
```
it 'filters products by category' do
    filter_by Category
  end

  it 'filters products by brand' do
    filter_by Brand
  end
``` 
...ale wydaje mi się że to nie najlepsze rozwiązanie bo żeby dowiedzieć się co się tak naprawdę dzieje rzeba sprawdzić helper.

Reply:
obczaj factory_girl i lety :v: